src/logics/syncswap/utils.ts

```diff
@@ -0,0 +1,688 @@
+  UINT256_MAX,
+  ZERO,
+} from './constants';
+import { IRouter } from './contracts/Router';
+import { RouteHelper } from './contracts';
+import { getToken } from './configs';
+
+function normalizePools(routePools: RouteHelper.RoutePoolStructOutput[]) {
+  return routePools.map((routePool) => ({
+    pool: routePool.pool,
+    tokenA: routePool.tokenA,
+    tokenB: routePool.tokenB,
+    poolType: routePool.poolType,
+    reserveA: routePool.reserveA,
+    reserveB: routePool.reserveB,
+    swapFeeAB: routePool.swapFeeAB,
+    swapFeeBA: routePool.swapFeeBA,
+  }));
+}
+
+export function normalizeRoutePools(routePools: RouteHelper.RoutePoolsStructOutput): RoutePools {
+  return {
+    poolsDirect: normalizePools(routePools.poolsDirect),
+    poolsA: normalizePools(routePools.poolsA),
+    poolsB: normalizePools(routePools.poolsB),
+    poolsBase: normalizePools(routePools.poolsBase),
+  };
+}
+
+function hasLiquidity(routePool: RoutePool, tokenIn?: string, amountOut?: BigNumber) {
+  if (tokenIn && amountOut) {
+    const reserveOut = routePool.tokenA === tokenIn ? routePool.reserveB : routePool.reserveA;
+    return reserveOut.gt(amountOut);
+  }
+
+  return routePool.reserveA.gte(LIQUIDITY_MIN_RESERVE) && routePool.reserveB.gte(LIQUIDITY_MIN_RESERVE);
+}
+
+function createPath(steps: Omit<Step, 'swapFee'>[]): Path {
+  return {
+    steps: steps.map(({ routePool, tokenIn }) => ({
+      routePool,
+      tokenIn,
+      swapFee: tokenIn === routePool.tokenA ? routePool.swapFeeAB : routePool.swapFeeBA,
+    })),
+  };
+}
+
+function newRoutePoolKey(tokenA: string, tokenB: string, poolType: number) {
+  return tokenA.toLowerCase() < tokenB.toLowerCase()
+    ? `${tokenA}:${tokenB}:${poolType}`
+    : `${tokenB}:${tokenA}:${poolType}`;
+}
+
+function getPathsWith1Hop(poolsA: RoutePool[], poolsB: RoutePool[], tokenIn: string, baseToken: string) {
+  const paths: Path[] = [];
+  for (const poolA of poolsA) {
+    for (const poolB of poolsB) {
+      if (poolA && poolB) {
+        paths.push(
+          createPath([
+            { routePool: poolA, tokenIn },
+            { routePool: poolB, tokenIn: baseToken },
+          ])
+        );
+      }
+    }
+  }
+
+  return paths;
+}
+
+function getPathsWith2Hops(
+  poolsA: RoutePool[],
+  poolsBase: RoutePool[],
+  poolsB: RoutePool[],
+  tokenIn: string,
+  baseToken1: string,
+  baseToken2: string
+) {
+  const paths: Path[] = [];
+  for (const poolA of poolsA) {
+    for (const poolBase of poolsBase) {
+      for (const poolB of poolsB) {
+        if (poolA && poolBase && poolB) {
+          paths.push(
+            createPath([
+              { routePool: poolA, tokenIn },
+              { routePool: poolBase, tokenIn: baseToken1 },
+              { routePool: poolB, tokenIn: baseToken2 },
+            ])
+          );
+        }
+      }
+    }
+  }
+
+  return paths;
+}
+
+export function findAllPossiblePaths(tokenIn: string, tokenOut: string, routePools: RoutePools, baseTokens: string[]) {
```

Comment:
No, it just the base tokens.
https://gist.github.com/0xnakato/9f202aaf09f3990a04697718e6076070#file-routepools_fetcher-ts-L25
https://gist.github.com/0xnakato/9f202aaf09f3990a04697718e6076070#file-routepools_fetcher-ts-L32
https://gist.github.com/0xnakato/9f202aaf09f3990a04697718e6076070#file-routepools_fetcher-ts-L41
https://gist.github.com/0xnakato/34bb1224099be58da312f154ba0ff9c4#file-routehelper-sol-L144
https://gist.github.com/0xnakato/9f202aaf09f3990a04697718e6076070#file-routepools_fetcher-ts-L57

Reply:
Hmm i only see that is comments `baseToken` after `routeTokens`, but not how it is used. But i guess it's ok to just give it a try since we should have other ways such as slippage to make sure that user gets enough tokens.